src/main/java/net/minecraftforge/unification/UnificationRegistry.java

```diff
@@ -0,0 +1,153 @@
+
+import com.google.common.base.Preconditions;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.ForgeModContainer;
+import net.minecraftforge.oredict.ShapedOreRecipe;
+import net.minecraftforge.oredict.ShapelessOreRecipe;
+
+/**
+ * The {@link UnificationRegistry} allows mods to register *basic* materials so they can be unified.
+ * <p>
+ * Using the {@link UnificationRegistry}, everything can return a single unified type,
+ * so for example players will only ever get one kind of Copper Ingot or Tin Ore.
+ * <p>
+ * Basic materials follow vanilla conventions, like iron ore being smeltable into iron ingots.
+ * If other mods cannot make basic assumptions about your materials, do not register them for unification.
+ * <p>
+ * <b>Usage:</b>
+ * <p>
+ * First, create your {@link UnificationMaterial}s and {@link UnificationForm}s or use the ones listed in {@link UnificationConstants}.
+ * <p>
+ * Then, add your mod's variant with {@link #add(UnificationMaterial, UnificationForm, T)}.
+ * You will get back instances of {@link Unified}, which is used to unify all things with those properties.
+ * <p>
+ * <b>Recipes:</b>
+ * <p>
+ * Recipes created with {@link Unified} for an ingredient will accept any item registered for a type, and return just the one unified type.
+ * {@link Unified} can be used in recipes like {@link ShapedOreRecipe} and {@link ShapelessOreRecipe},
+ * or you can implement your own recipes.
+ * <p>
+ * You can also create JSON recipes that use {@link Unified} as an ingredient or result, by using
+ * the recipe type "forge:ore_shaped" or "forge:ore_shapeless", and using the ingredient format
+ * {"type":"forge:unified", "material": "materialUid", "form": "formUid"}
+ * For examples see the json recipes in the forge test directory "src\test\resources\assets\item_unification_test\recipes"
+ */
+public final class UnificationRegistry<T>
+{
+    private final Map<UnificationMaterial, Map<UnificationForm, Unified<T>>> UNIFICATION_TABLE = new IdentityHashMap<>();
+    private final Comparator<T> sortOrder;
+    private final Function<T, T> validator;
+    private final Function<T, T> outputCopier;
+
+    /**
+     * Create a new instance with {@link UnificationManager#createRegistry(Function, Function, Function)}
+     */
+    UnificationRegistry(Function<T, ResourceLocation> namer, Function<T, T> validator, Function<T, T> outputCopier)
+    {
+        this.sortOrder = new UnifiedSortOrder<>(namer);
+        this.validator = validator;
+        this.outputCopier = outputCopier;
+    }
+
+    /**
+     * Add a variant to the registry.
+     * If no {@link Unified} exists yet, one is created.
+     * If a {@link Unified} already exist with these properties, the variant is added to its list of variants.
+     * For example usage, see {@link UnificationConstants#initVanillaEntries()}
+     */
+    public Unified<T> add(UnificationMaterial material, UnificationForm shape, T variant)
```

Comment:
`addIfAbsent`?

Reply:
This adds a variant to the `Unified` entry (even if one already exists). They are used for some things, like recipes will accept any of the registered variants.